File: copy_list_with_random_pointer/CopyListWithRandomPointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

public class CopyListWithRandomPointer {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        return copyRandomList(head, new HashMap<>());
    }
    
    private Node copyRandomList(Node head, Map<Node, Node> nodes) {
        if (head == null) {
            return null;
        }
        if (nodes.containsKey(head)) {
            return nodes.get(head);
        }
        if (!nodes.containsKey(head)) {
            nodes.put(head, new Node(head.val));
        }
        
        Node newHead = nodes.get(head);
        newHead.next = copyRandomList(head.next, nodes);
        newHead.random = copyRandomList(head.random, nodes);
        return newHead;
    }
}

File: copy_list_with_random_pointer/Copy_List_With_Random_Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Copy_List_With_Random_Pointer {
public:
    Node* copyRandomList(Node* head) {
        if (head == NULL) {
            return NULL;
        }
        
        unordered_map<Node*, Node*> nodes;
        return dfs(head, nodes);
    }
private: 
    Node* dfs(Node* node, unordered_map<Node*, Node*>& nodes) {
        if (node == NULL) {
            return NULL;
        }
        if (nodes.find(node) != nodes.end()) {
            return nodes[node];
        }

        nodes[node] = new Node(node->val);
        nodes[node]->next = dfs(node->next, nodes);
        nodes[node]->random = dfs(node->random, nodes);
        return nodes[node];
    }
};